### PR TITLE
Moved backend code to new package 'chat'. Also added test.

### PR DESCRIPTION
**Pull request by GoldenBadger**
Please be very very critical. I know you're not too familiar with Go, so just ask lots of questions. During last year's projects everyone either automatically accepted pull requests or merged their own ones, and that was a _really bad thing_.

### CODE REVIEW

**Comment by GoldenBadger** [comment]
THANK GOD. Basically the problem was that Travis clones the repo with mixed case while go import paths are always lower case. Therefore, Travis had no idea what the hell was going on.

File: chat/chat.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	Register chan *Connection
+
+	// Requests to leave the room
+	Unregister chan *Connection
+}
+
+func NewRoom() *Room {
+	return &Room{
+		connections: make(map[*Connection]bool),
+		messages:    make(chan []byte, 20),
```
Comment (VoyTechnology) [inline chat/chat.go:31]: Can you explain this line?

File: chat/chat.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	return &Connection{
+		ws:   ws,
+		r:    r,
+		mess: make(chan []byte, 256),
+	}
+}
+
+func (conn *Connection) Reader() {
+	for {
+		_, message, err := conn.ws.ReadMessage()
```
Comment (VoyTechnology) [inline chat/chat.go:72]: What is _ ?

**Comment by VoyTechnology** [comment]
Tbh I haven't actually worked with pure Websockets, but looking at [MDN's ](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_client_applications) description, what do we put for protocol, if needed?

File: chat/chat.go (hunk 1 above)
Comment (GoldenBadger) [inline chat/chat.go:31]: For line 31 it means create a channel (basically a queue) buffered to hold 20 byte arrays before blocking. I should change it to 256 actually, to keep it consistent with the others.

File: chat/chat.go (hunk 2 above)
Comment (GoldenBadger) [inline chat/chat.go:72]: In Go, functions can return multiple things. This one returns three, and I only care about the last 2, so "_" is used to ignore the first result.

File: chat/chat.go (hunk 1 above)
Comment (VoyTechnology) [inline chat/chat.go:31]: why 20 bytes?

**Comment by GoldenBadger** [comment]
You mean in the URL? It'll look like `ws://localhost:8080/` or whatever.

**Comment by VoyTechnology** [comment]
I know about the url, but if you look at

```
var exampleSocket = new WebSocket("ws://www.example.com/socketserver", "protocolOne");
```

I am intersted about the `protocolOne` bit

File: chat/chat.go (hunk 1 above)
Comment (GoldenBadger) [inline chat/chat.go:31]: Not an array of 20 bytes. 20 byte-arrays of any size.

**Comment by GoldenBadger** [comment]
I'm not sure. That's more of a client-side thing, from what I can tell. See [here](https://stackoverflow.com/questions/7363095/javascript-and-websockets-using-specific-protocol).

**Pull request merged**